### PR TITLE
New integration test framework for security

### PR DESCRIPTION
**Pull request by mangchiandjjoe**
**What this PR does / why we need it**:

Integration framework dogfood integration for security

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

This is the rewriting the /security/integration test

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by mangchiandjjoe** [comment]
/assign @wattli @yutongz @hklai 
PTAL

**Comment by wattli** [comment]
@rshriram , i agree there are some dups between security and pilot. However, I think the idea of the new integration framework is to provide a unified platform such that each components can use separately. Besides, some of the logic does not quite make sense to be added into pilot, e.g., the communication between node agent and CA for CSR, which is via direct gRPC. And the communication between node agent and workloads. Moreover, there will be authorization test cases soon in security.

In short, i would prefer to have this security integration dir. And have some test cases covered in pilot.

File: security/tests/integration/cmd/main.go [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+	"istio.io/istio/tests/integration/framework"
+)
+
+const (
+	testID = "istio_ca_secret_test"
+	// Specifies how long we wait before a secret becomes existent.
+	secretWaitTime = 20 * time.Second
+	// Certificates validation retry
+	certValidateRetry      = 10
+	certValidationInterval = 1 // Initially wait for 1 second. This value will be increased exponentially on retry
```
Comment (wattli) [inline security/tests/integration/cmd/main.go]: What about putting the trailing comment on top to be consistent with previous ones?

File: security/tests/integration/istio_ca_secret_env.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+				"--self-signed-ca",
+			},
+		),
+	}
+}
+
+// Bringup doing general setup for environment level, not components.
+// Bringup() is called by framework.SetUp()
+func (env *SecretTestEnv) Bringup() error {
+
```
Comment (wattli) [inline security/tests/integration/istio_ca_secret_env.go]: Remove empty line. Do we actually need this a dummy function here?

**Comment by kyessenov** [comment]
Yes, use the right tool for the job. I also don't think we necessarily need
to reuse pilot's test framework for security.

File: security/tests/integration/kubernetes_service.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+		map[string]string{
+			"uuid": c.uuid,
+		},
+		c.selector,
+		c.annotation)
+	if err != nil {
+		return fmt.Errorf("failed to deploy Istio CA (error: %v)", err)
+	}
+
+	return err
```
Comment (wattli) [inline security/tests/integration/kubernetes_service.go]: The logic seems strange, here err is nil?

**Comment by wattli** [comment]
And we will remove the existing integration test, right?

File: security/tests/integration/istio_ca_secret_env.go (hunk 2 above)
Comment (mangchiandjjoe) [reply security/tests/integration/istio_ca_secret_env.go]: It needs to be implemented. Removed the empty line

File: security/tests/integration/kubernetes_service.go (hunk 3 above)
Comment (mangchiandjjoe) [reply security/tests/integration/kubernetes_service.go]: err should be nil. fixed err handling.

File: security/tests/integration/kubernetes_utils.go [hunk 4]
```diff
@@ -0,0 +1,371 @@
+// Copyright 2017 Istio Authors
```
Comment (yutongz) [inline security/tests/integration/kubernetes_utils.go]: It would be good to move this util library into tests/util

File: security/tests/integration/cmd/main.go [hunk 5]
```diff
@@ -0,0 +1,276 @@
+// Copyright 2017 Istio Authors
```
Comment (yutongz) [inline security/tests/integration/cmd/main.go]: Ideally we should use go test for this instead a go binary. Just like sample tests.

File: security/tests/integration/cmd/main.go [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+	certChain := flag.String("cert-chain", "", "Path to the original workload certificate chain")
+
+	flag.Parse()
+
+	clientset, err := integration.CreateClientset(*kubeconfig)
+	if err != nil {
+		glog.Fatalf("failed to initialize K8s client: %s\n", err)
+	}
+
+	env := integration.NewSecretTestEnv("Secret creation and recovery test", clientset, *hub, *tag)
```
Comment (yutongz) [inline security/tests/integration/cmd/main.go]: This part can be handled by testEM.runtest() if you are using go test.  https://github.com/istio/istio/blob/master/tests/integration/example/tests/sample2/sample2_test.go#L63

File: security/tests/integration/cmd/main.go [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+
+		return nil
+	}
+
+	return fmt.Errorf("failed to check certificate update and validate after %v retry", certValidateRetry)
+}
+
+func main() {
+	kubeconfig := flag.String("kube-config", "", "path to kubeconfig file")
+	hub := flag.String("hub", "", "Docker hub that the Istio CA image is hosted")
```
Comment (yutongz) [inline security/tests/integration/cmd/main.go]: For me, it makes more sense to put flag `hub` and `tag` down to component level

File: security/tests/integration/istio_ca_secret_env.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+// GetComponents is the key of a environment
+// It defines what components a environment contains.
+// Components will be stored in framework for start and stop
+func (env *SecretTestEnv) GetComponents() []framework.Component {
+	return []framework.Component{
+		NewKubernetesPod(
+			env.ClientSet,
+			env.NameSpace,
+			"istio-ca-self-signed",
```
Comment (yutongz) [inline security/tests/integration/istio_ca_secret_env.go]: Can you make it const

**Comment by rshriram** [comment]
if you are doing all this, can you please add a script or whatever necessary to run the same test on circleci which is using a pure go build environment. Or if you make it run in presubmit, thats fine as well. You can take a look at pilot-integration-test-auth to see how I am invoking integration tests without bazel

**Comment by wattli** [comment]
@rshriram , as discussed offline, currently we are switching to a new integration testing framework initiated by @yutongz and @hklai , here is the original PR (docs can be found there too): https://github.com/istio/istio/pull/1936

And I think the scope of this PR is simply to switching to this new framework and we can have follow-up PRs for build tools (bazel vs non-bazel) once it is finalized.

**Comment by mangchiandjjoe** [comment]
@rshriram unittest-security is getting failed because of tests added in the PR requires arguments.
Is there a way I can exclude new tests?

**Comment by rshriram** [comment]
Merge with master please. can you add the args to unit test as well? These are standard go tests. It’s in .circleci/config.yml ..

File: security/tests/integration/certificateRotationTestEnv/integration_test.go [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+	config = &Config{
+		kubeconfig: *kubeconfig,
+		hub:        *hub,
+		tag:        *tag,
+	}
+
+	glog.Errorf("%v", config)
+
+	clientset, err := integration.CreateClientset(config.kubeconfig)
```
Comment (yutongz) [inline security/tests/integration/certificateRotationTestEnv/integration_test.go]: Can we move it into the env?

File: security/tests/integration/certificateRotationTestEnv/integration_test.go [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+		glog.Fatalf("failed to initialize K8s client: %s\n", err)
+	}
+
+	testEnv = integration.NewCertRotationTestEnv(testEnvName, clientset, config.hub, config.tag)
+
+	res := framework.NewTestEnvManager(testEnv, testID).RunTest(m)
+
+	glog.Infof("Test result %d in env %s", res, testEnvName)
+
+	if res == 0 {
```
Comment (yutongz) [inline security/tests/integration/certificateRotationTestEnv/integration_test.go]: os.Exit(res)

File: security/tests/integration/nodeAgentTestEnv/integration_test.go [hunk 11]
```diff
@@ -0,0 +1,191 @@
+// Copyright 2017 Istio Authors
```
Comment (yutongz) [inline security/tests/integration/nodeAgentTestEnv/integration_test.go]: if you name the folder nodeAgentTest**Env**, then it makes more sense to put _env.go under this folder.

Comment (mangchiandjjoe) [reply security/tests/integration/nodeAgentTestEnv/integration_test.go]: Renamed directories and files

File: security/tests/integration/certificateRotationTestEnv/integration_test.go (hunk 10 above)
Comment (mangchiandjjoe) [reply security/tests/integration/certificateRotationTestEnv/integration_test.go]: Applied

File: security/tests/integration/certificateRotationTestEnv/integration_test.go (hunk 9 above)
Comment (mangchiandjjoe) [reply security/tests/integration/certificateRotationTestEnv/integration_test.go]: Moved

File: security/tests/integration/cmd/main.go (hunk 7 above)
Comment (yutongz) [reply security/tests/integration/cmd/main.go]: Thoughts?

File: security/tests/integration/kubernetes_utils.go (hunk 4 above)
Comment (yutongz) [reply security/tests/integration/kubernetes_utils.go]: Do you think it makes sense, or maybe we can do it later :)

File: security/tests/integration/cmd/main.go (hunk 7 above)
Comment (mangchiandjjoe) [reply security/tests/integration/cmd/main.go]: main.go is deprecated.

File: security/tests/integration/kubernetes_utils.go (hunk 4 above)
Comment (mangchiandjjoe) [reply security/tests/integration/kubernetes_utils.go]: pilot has a K8s util frameworks for the integration. I want to evaluate the library first.
we can decide it later.

File: security/tests/integration/cmd/main.go (hunk 7 above)
Comment (yutongz) [reply security/tests/integration/cmd/main.go]: Right, but we still keep these flags in testMain(). I think it makes more sense logically to move `hub` and `tag` to component level and kubeconfig can be moved to env or so :)

File: security/tests/integration/istio_ca_node_agent_env.go [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+	glog.Infof("cleaning up environment...")
+	err := deleteTestNamespace(env.ClientSet, env.NameSpace)
+	if err != nil {
+		glog.Errorf("failed to delete namespace: %v error: %v", env.NameSpace, err)
+	}
+	return nil
+}
+
+// GetNodeAgentIPAddress returns the external LoadBalancer IP address of the service
+func (env *NodeAgentTestEnv) GetNodeAgentIPAddress() (string, error) {
```
Comment (hklai) [inline security/tests/integration/istio_ca_node_agent_env.go]: @yutongz 

We should add some discovery support in the integration test framework so people can find the component addresses from an environment in a uniform way.

File: security/tests/integration/kubernetes_pod.go [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+	"fmt"
+
+	"github.com/golang/glog"
+	"istio.io/istio/tests/integration/framework"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/client-go/kubernetes"
+)
+
+type (
+	// KubernetesPod is the test component for K8s pod
```
Comment (hklai) [inline security/tests/integration/kubernetes_pod.go]: Can these components be reused by other tests outside of security?
If so, should these be in the common component directory?

File: security/tests/integration/kubernetes_service.go [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+// Stop stop this component
+// Stop is being called in framework.TearDown()
+func (c *KubernetesService) Stop() (err error) {
+	glog.Infof("deleting the service: %v", c.name)
+	return deleteService(c.clientset, c.namespace, c.name)
```
Comment (hklai) [inline security/tests/integration/kubernetes_service.go]: It seems like all/most Stop() cleans up the underlying resources instead of just deleting.
Is it intended or is it because of some limitations?

@yutongz, this is actually asymmetrical API. There is no separate Create() method. Is CleanUp() meant to be used as destructor?

File: security/tests/integration/istio_ca_node_agent_env.go (hunk 12 above)
Comment (yutongz) [reply security/tests/integration/istio_ca_node_agent_env.go]: Agree. I think we can add the `status` concept into framework and component address can be a field.

File: security/tests/integration/kubernetes_service.go (hunk 14 above)
Comment (yutongz) [reply security/tests/integration/kubernetes_service.go]: I think usually golang doesn't require a destructor. And thus we can probably merge stop() and cleanup().

**Comment by yutongz** [comment]
Curious where is this script `security/bin/e2e.sh` being called in Prow cycle?

File: prow/istio-presubmit.sh [hunk 15]
```diff
@@ -113,5 +113,8 @@ if [[ -n $(git diff) ]]; then
   git diff
 fi
 
+# run security e2e test
+${ROOT}/security/bin/e2e.sh --hub "gcr.io/istio-testing" --tag "${GIT_SHA}"
```
Comment (yutongz) [inline prow/istio-presubmit.sh]: I said it wrong, we need to put this after `make push`

File: security/tests/integration/istio_ca_cert_rotation_env.go [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+// GetComponents is the key of a environment
+// It defines what components a environment contains.
+// Components will be stored in framework for start and stop
+func (env *CertRotationTestEnv) GetComponents() []framework.Component {
+	return []framework.Component{
+		NewKubernetesPod(
+			env.ClientSet,
+			env.NameSpace,
+			istioCaSelfSignedShortTTL,
+			fmt.Sprintf("%v/istio-ca:%v", env.Hub, env.Tag),
```
Comment (yutongz) [inline security/tests/integration/istio_ca_cert_rotation_env.go]: How about we put `hub` and `tag` directly into `kubernetes_pod.go`

File: prow/istio-presubmit.sh (hunk 15 above)
Comment (mangchiandjjoe) [reply prow/istio-presubmit.sh]: Done.

File: security/tests/integration/istio_ca_cert_rotation_env.go (hunk 16 above)
Comment (mangchiandjjoe) [reply security/tests/integration/istio_ca_cert_rotation_env.go]: Done

**Comment by yutongz** [comment]
Awesome job Jae! 
/lgtm

Do you have anything else to add @mangchiandjjoe,  otherwise I think it's good to merge @hklai @wattli

File: security/tests/integration/istio_ca_node_agent_env.go [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+			env.NameSpace,
+			nodeAgent,
+			fmt.Sprintf("%v/node-agent-test:%v", env.Hub, env.Tag),
+			[]string{},
+			[]string{},
+		),
+		NewKubernetesService(
+			env.ClientSet,
+			env.NameSpace,
+			nodeAgentService,
```
[the remark this answers is not in the document]
Comment (mangchiandjjoe) [reply security/tests/integration/istio_ca_node_agent_env.go]: The NodeAgent service is for the web service written in JavaScript to download certificates from the pod.

[the remark this answers is not in the document]
Comment (myidpt) [reply security/tests/integration/istio_ca_node_agent_env.go]: I see. Thanks for clarifying.

File: security/tests/integration/istio_ca_cert_rotation_env.go [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+				"--cert-ttl", "60s",
+			},
+		),
+	}
+}
+
+// Bringup doing general setup for environment level, not components.
+// Bringup() is called by framework.SetUp()
+func (env *CertRotationTestEnv) Bringup() error {
+	return nil
```
Comment (hklai) [inline security/tests/integration/istio_ca_cert_rotation_env.go]: @yutongz Probably not for this PR, but I think it looks more natural to have BringUp() starts all components internally. That may not be possible since the test Env does not 
really inherit from a base/common environment impl.

Currently it looks redundant to the constructor/new method. Perhaps we can just get rid of it?

Comment (yutongz) [reply security/tests/integration/istio_ca_cert_rotation_env.go]: I agree. I will fix this in my improvement PR.
